src/util/BinaryFuseFilter.h

```diff
@@ -0,0 +1,91 @@
+#pragma once
+
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "lib/binaryfusefilter.h"
+#include "util/NonCopyable.h"
+#include "xdr/Stellar-ledger-entries.h"
+#include "xdr/Stellar-types.h"
+
+#include <cereal/archives/binary.hpp>
+#include <xdrpp/xdrpp/cereal.h>
+
+namespace stellar
+{
+
+// This class is a wrapper around the binary_fuse_t library that provides
+// serialization for the XDR BinaryFuseFilter type and provides a deterministic
+// LedgerKey interface.
+
+// Only allow uint8_t, uint16_t, and uint32_t
+template <typename T,
+          typename = std::enable_if_t<std::is_unsigned<T>::value &&
+                                      !std::is_same<T, uint64_t>::value>>
+class BinaryFuseFilter : public NonMovableOrCopyable
+{
+  private:
+    binary_fuse_t<T> mFilter;
+
+    // Note: as part of filter construction, the internal filter seed might
+    // rotate and no longer be the same as the input seed. The input seed must
+    // be maintained outside of the filter and used to hash input keys in the
+    // contain function to ensure deterministic hashing of input keys
+    // during both populating and querying the filter.
+    binary_fuse_seed_t const mInputSeed;
+
+  public:
+    // keyHashes is the SipHash24 digest on the keys to insert into the filter.
+    // Seed is the random seed used to initialize the hash function
+    explicit BinaryFuseFilter(std::vector<uint64_t>& keyHashes,
+                              binary_fuse_seed_t const& seed);
+    explicit BinaryFuseFilter(SerializedBinaryFuseFilter const& xdrFilter);
+
+    bool contains(LedgerKey const& key) const;
+
+    bool operator==(BinaryFuseFilter<T> const& other) const;
+
+    template <class Archive>
+    void
+    save(Archive& archive) const
```

Comment:
Why do we need to use custom encoding here instead of copying the fields into the XDR struct?

Reply:
Hmm I'm not sure what you mean by custom encoding? We are just copying the fields into the XDR struct. The `mInputSeed` parameter is used to provide input into the binary fuse filter library, so it is stored in `mInputSeed` instead of the library class `mFilter`. `mFilter.copyTo` does just copy all the fields into the XDR struct.

If by "custom encoding" you're refering to byte packing the `fingerprints` XDR field, that is necessary because XDR does not support vector with 16 bit elements, only 8 bit and 32 bit. For consistency I opted to always use an 8 bit XDR vector and bit pack all elem types instead of just special casing the 16 bit case and requiring a switch at the XDR level.

Reply:
Yeah, by 'custom encoding' I meant definitions of save/load functions instead of only using the autogenerated ones for the XDR struct.  I'm still not sure I understand the motivation for implementing the archive functions vs explicit construction from/conversion to `SerializedBinaryFuseFilter`; typically we use that to customize the way the data is being serialized, but IIUC we actually don't customize anything. But I'm not feeling too strongly about this.

Reply:
This is just for convenience purposes. After start and when creating a new filter, we only want to keep the non-XDR version in memory to actually do lookups/construction. When we go to serialize, since we only have the in-memory version, it's significantly simpler to just define a skeleton save/load pair here than manually convert the saved version to the XDR version then serialize that.

Specifically, in `SerializableBucketIndex`, we want to serialize `std::unique_ptr<BinaryFuseFilter16> filter`. Cereal has support for smart pointers, where it essentially automatically does
```
ar(pointerOverhead)
if (pointer)
    ar(*pointer)
```

If we were to not specify thesave/load pairs for the non-xdr filter, we'd have to manually implement this pointer behavior instead of just using the built in cereal functionality, since Cereal would not be able to deduce how to archive the value of the dereferenced pointer.